bindings/wasm/output.go

```diff
@@ -0,0 +1,192 @@
+	"bytes"
+	"context"
+	"crypto/rand"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+	"sync/atomic"
+
+	"github.com/tetratelabs/wazero"
+	"github.com/tetratelabs/wazero/api"
+	"github.com/tetratelabs/wazero/imports/wasi_snapshot_preview1"
+	"github.com/tetratelabs/wazero/sys"
+
+	"github.com/dapr/components-contrib/bindings"
+	"github.com/dapr/kit/logger"
+)
+
+type initMetadata struct {
+	// Path is where to load a `%.wasm` file that implements a command,
+	// usually compiled to target WASI.
+	Path string `json:"path"`
+
+	// guest is WebAssembly binary implementing the waPC guest, loaded from Path.
+	guest []byte
+}
+
+type outputBinding struct {
+	logger        logger.Logger
+	runtimeConfig wazero.RuntimeConfig
+	moduleConfig  wazero.ModuleConfig
+
+	runtime wazero.Runtime
+	module  wazero.CompiledModule
+
+	instanceCounter uint64
+}
+
+var (
+	_ bindings.OutputBinding = (*outputBinding)(nil)
+	_ io.Closer              = (*outputBinding)(nil)
+)
+
+func NewWasmOutput(logger logger.Logger) bindings.OutputBinding {
+	return &outputBinding{
+		logger: logger,
+
+		// The below ensures context cancels in-flight wasm functions.
+		runtimeConfig: wazero.NewRuntimeConfig(),
+
+		// The below violate sand-boxing, but allow code to behave as expecteout.
+		moduleConfig: wazero.NewModuleConfig().
+			WithRandSource(rand.Reader).
+			WithSysNanosleep().
+			WithSysWalltime().
+			WithSysNanosleep(),
+	}
+}
+
+func (out *outputBinding) Init(ctx context.Context, metadata bindings.Metadata) (err error) {
+	meta, err := out.getInitMetadata(metadata)
+	if err != nil {
+		return fmt.Errorf("wasm: failed to parse metadata: %w", err)
+	}
+
+	out.runtime = wazero.NewRuntimeWithConfig(ctx, out.runtimeConfig)
+	out.module, err = out.runtime.CompileModule(ctx, meta.guest)
+	if err != nil {
+		_ = out.runtime.Close(ctx)
+		return fmt.Errorf("wasm: error compiling binary: %w", err)
+	}
+
+	switch detectImports(out.module.ImportedFunctions()) {
+	case modeWasiP1:
+		_, err = wasi_snapshot_preview1.Instantiate(ctx, out.runtime)
+	}
+
+	if err != nil {
+		return fmt.Errorf("wasm: error instantiating host functions: %w", err)
+	}
+	return
+}
+
+func (out *outputBinding) Invoke(ctx context.Context, req *bindings.InvokeRequest) (*bindings.InvokeResponse, error) {
+	// Currently, concurrent modules can conflict on name. Make sure we have
+	// a unique one.
+	moduleName := strconv.FormatUint(atomic.AddUint64(&out.instanceCounter, 1), 10)
+	moduleConfig := out.moduleConfig.WithName(moduleName)
+
+	if len(req.Data) > 0 {
+		moduleConfig = moduleConfig.WithStdin(bytes.NewReader(req.Data))
+	}
+
+	var stdout bytes.Buffer
+	moduleConfig = moduleConfig.WithStdout(&stdout)
+
+	if args, ok := req.Metadata["args"]; ok {
+		argsSlice := strings.SplitN(args, ",", -1)
+		// prepend arg0 (program name) unless we want users to supply it.
+		argsSlice = append([]string{out.module.Name()}, argsSlice...)
+		moduleConfig = moduleConfig.WithArgs(argsSlice...)
```

Comment:
If `req.Metadata["args"]` is not defined, you don't want to pass `out.module.Name()` as args?

Reply:
so arg0 is the program name.. it is a bit arbitrary what we choose here, but some things expect that even if there are no args... wdyt?

Reply:
I don't have a problem with arg0.

What I was trying to say is that you set `moduleConfig = moduleConfig.WithArgs(argsSlice...)` only within the block with `if req.Metadata["args"]`. If `req.Metadata["args"]` is empty, you don't want to set `moduleConfig.WithArgs`?

Reply:
made it more explicit, and tested it. please re-open this if the direction isn't correct.